server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridAggregatorFactory.java

```diff
@@ -79,15 +83,26 @@ protected Aggregator doCreateInternal(final ValuesSource valuesSource,
                                             boolean collectsFromSingleBucket,
                                             List<PipelineAggregator> pipelineAggregators,
                                             Map<String, Object> metaData) throws IOException {
-        if (valuesSource instanceof ValuesSource.GeoPoint  == false) {
-            throw new AggregationExecutionException("ValuesSource type " + valuesSource.toString() + "is not supported for aggregation " +
-                this.name());
+        AggregatorSupplier aggregatorSupplier = queryShardContext.getValuesSourceRegistry()
+            .getAggregator(config.valueSourceType(), GeoHashGridAggregationBuilder.NAME);
+        if (aggregatorSupplier instanceof GeoGridAggregatorSupplier == false) {
+            throw new AggregationExecutionException("Registry miss-match - expected "
+                + GeoGridAggregatorSupplier.class.getName() + ", found [" + aggregatorSupplier.getClass().toString() + "]");
         }
         if (collectsFromSingleBucket == false) {
             return asMultiBucketAggregator(this, searchContext, parent);
         }
         CellIdSource cellIdSource = new CellIdSource((ValuesSource.GeoPoint) valuesSource, precision, geoBoundingBox, Geohash::longEncode);
```

Comment:
I think we need to move creating the `CellIdSource` into the supplier.  This line is casting to a `GeoPoint`, but we don't know that we always have a `GeoPoint` here, because we can't know at compile time what Values Sources have been registered for this aggregation.  My guideline has been that we should only care about `ValuesSource` subclasses at the supplier layer or the aggregator itself, not the factory or the builder.

Reply:
++ will do